openedx/openedx-platform
# [BD-13] Deprecate ModuleSystem.render_template (fixed)

## Description

Re-submits the changes from https://github.com/edx/edx-platform/pull/28571 which were reverted by https://github.com/edx/edx-platform/pull/29341, and fixes the bug that caused this revert.

Builds on the shim added by https://github.com/edx/edx-platform/pull/28440 to deprecate the `ModuleSystem.render_template` attribute, in favor of using the added `MakoService.render_template`.

This change is only a refactoring, and should not affect Learners, Course Authors, or anyone else using edx-platform.

## Supporting information

See https://github.com/edx/edx-platform/pull/28571#issuecomment-971893521 for details about why that PR was reverted.  343a182 adds a test for this issue, demonstrating that the [new test fails](https://build.testeng.edx.org/job/edx-platform-python-pipeline-pr/36743/), but is fixed by the subsequent commt 8c874c2.

See also:

* [Content Core Platform Simplification project](https://openedx.atlassian.net/wiki/spaces/AC/pages/3091629503/Content+Core+Platform+Simplification)
* [BLENDED-113](https://openedx.atlassian.net/browse/BLENDED-113) (edX internal ticket)
* [FAL-2304](https://tasks.opencraft.com/browse/FAL-2304). [FAL-2603](https://tasks.opencraft.com/browse/FAL-2603) (OpenCraft internal tickets)

## Testing instructions

Sandbox provisioning:

* LMS: https://pr29354.sandbox.opencraft.hosting/
* Studio: https://studio.pr29354.sandbox.opencraft.hosting/

LMS as learner:
1. [Login](https://pr29354.sandbox.opencraft.hosting/login) as a demo learner (`honor@example.com` / `edx`)
2. Browse around the [BD-13 test course](https://discovery.pr29354.sandbox.opencraft.hosting/learning/course/course-v1:OpenCraft+BD-13+2021_1/home) and ensure that blocks render as expected, and that you can submit answers to problems. This course has all of the affected XBlocks in it, including an example Video block which shows the CC license wrapper (cf 7fd14678c2f3b4f772e3ae27b254082a34bfee7e).
3. Switch to the Legacy Course Experience and browse around in that too.

LMS as course staff:
1. [Login](https://pr29354.sandbox.opencraft.hosting/login) as a demo staff user (`staff@example.com` / `edx`)
2. Navigate to a course and its Instructor Dashboard > Email tab
3. Ensure that you can see and use the HTML WYSIWYG editor to craft emails to students. (cf 96d92083748145aa41001d85a28f293b338239ff)

Studio:
1. [Login](https://pr29354.sandbox.opencraft.hosting/login) as a demo staff (`staff@example.com` / `edx`).
2. Visit the [BD-13 test course in Studio](https://studio.pr29354.sandbox.opencraft.hosting/course/course-v1:OpenCraft+BD-13+2021_1) and browse through the blocks. Ensure they render and edit as expected.
3. Check that you can submit answers to problems, and save changes to blocks.
4. Preview unpublished changes and ensure they render as expected.
4. Publish changes, and ensure they render in the LMS as expected.

Test the Studio visibility bug:

This bug occurs only in the XBlocks which do not `@need("mako")`, like ORA2.

For OpenCraft staff:

* [Appserver 2](https://manage.opencraft.com/instance/31740/edx-appserver/21363/) contains the broken code, so you can verify the bug there.
* [Appserver 3](https://manage.opencraft.com/instance/31740/edx-appserver/21364/) is running the fixed code on this branch, so you can activate that appserver to verify the bug is fixed.

1. Login to Studio sandbox as a demo staff (staff@example.com / `edx`)
2. Navigate to the [BD-13 test course](https://studio.pr29354.sandbox.opencraft.hosting/course/course-v1:OpenCraft+BD-13+2021_1) and locate the Complex blocks > [ORA2 block](https://studio.pr29354.sandbox.opencraft.hosting/container/block-v1:OpenCraft+BD-13+2021_1+type@vertical+block@f9d578e6d981432a93e70b10924258d4).
3. Click on the "cog" icon to edit the block visibility.
    If verifying the bug exists: note that an error is shown in studio, saying "Error: service 'mako' was not requested"
    If verifying that the bug is fixed: ensure that you can edit the block's visibility.

## Deadline

None
